[2단계 - 로또] 이지윤 미션 제출합니다.

dd-jiyun opened the pull request:
안녕하세요 주노 🍀

이번 2단계는 1단계보다 뭔가 난이도가 좀 올라간 거 같아요. 
목표로 했던 `객체로 바라보며 설계`하였는지 걱정이네요 🥺

잘 부탁드립니다!

---

### 🤔 고민했던 부분
#### 당첨 번호 입력 구분 책임

입력 파싱과 구분 책임을 어디에 두는 것이 좋을지 항상 고민해요..
`1,2,3,4,5,6`과 같이 콤마(,)로 구분된 당첨 번호를 처리할 때, 파싱 전용 객체를 두기보다는 입력 문자열을 스스로 처리해 객체를 생성하는 것이 더 자연스러운 흐름이라 판단하여, 이번에는 당첨 번호 객체 내에서 파싱을 담당하도록 했어요.

#### 입출력 메세지 상수화
이번에 통계를 출력하게 되면서 `메세지를 상수화하는게 과연 가독성 있을까?`하는 고민이 생겼어요.
오히려 한 메서드에 모아두는 것이 어떤 순서로 출력이 될 것인지를 이해하기 쉬울 수 있을 것이라고 판단하였고, 어떤 기준으로 상수화를 진행하는게 좋을 지 생각해보게 되었습니다.

저의 상수화 기준은 아래와 같습니다!
- 변경 가능성이 있는 경우
- 클래스 내에서 여러 번 사용되는 경우
- 메시지 내용을 보고 역할이 명확히 드러나지 않는 경우 (예: 특정 조건을 알기 어려운 메시지)

### 🧐 궁금한 부분
- 통계 출력 시 메세지를 상수화하는 게 더 가독성이 높을 것이라고 생각하시나요?? 주노의 상수화 기준도 궁금합니다!
- 단계가 올라갈 때마다 더 객체를 중심으로 바라보려 노력해요. 지켜지지 않은 부분이 있는지, `이 부분은 다른 객체가 담당하는 게 자연스러운데?` 하고 생각이 드는 부분이 있다면 조언 부탁드립니다!

dd-jiyun committed:
docs: step2 기능 구현 목록 수정

dd-jiyun committed:
feat: 당첨 번호 객체 생성

- 입력받은 지난 당첨 번호를 가지고 있는 객체입니다.

dd-jiyun committed:
feat: 당첨 통계를 계산하는 객체 생성

dd-jiyun committed:
feat: 총 수익률 계산하는 객체 생성

dd-jiyun committed:
feat: 지난 당첨 번호 입력과 당첨 통계 및 수익률 출력 기능

dd-jiyun committed:
docs: 실행 결과 업데이트

dd-jiyun commented on src/main/java/view/InputView.java:
```diff
@@ -13,4 +13,10 @@ public static int inputPurchaseAmount() {
         System.out.println("구입금액을 입력해 주세요.(1000원 단위)");
         return Integer.parseInt(scanner.nextLine());
     }
+
+    public static String inputWinningNumberForLastWeek() {
+        System.out.println();
+        System.out.println("지난 주 당첨 번호를 입력해주세요.");
+        return scanner.nextLine();
```
### 🤩 알게된 내용
위의 `inputPurchaseAmount()`에서 문득 int로 반환하니 `nextInt`를 쓰면 되는데 왜 `parseInt`를 썼지 하고 바꿨어요.

그랬더니 구입 금액 입력하고, `inputWinningNumberForLastWeek()`로 지난 주 당첨 번호를 입력하려니까 빈 값에 설정해둔 에러가 터지더라고요.

이유가 무엇인지 찾아보니, `nextInt()`에서 `nextLine()`으로 호출하면 `nextInt()` 이후의 개행문자(\n)가 버퍼에 남아서 바로 빈 문자열로 읽어버린다는 사실을 알게 되었어요!

Choi-JJunho replied:
개행문자 버퍼문제를 겪으셨군요!
말씀해주신 내용처럼 커맨드라인에서 input을 받는 경우에는 개행문자를 고려하는 경우가 많아서 nextLine()을 자주 쓰는 편인것 같아요 👍

Choi-JJunho commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,34 @@
+
+import java.util.Arrays;
+
+public enum Rank {
+    FIRST(6, 2_000_000_000),
+    SECOND(5, 1_500_000),
+    THIRD(4, 50_000),
+    FOURTH(3, 5_000),
+    NONE(0, 0);
```
Rank를 Enum으로 표기해주셨군요 👍

Choi-JJunho commented on src/main/java/domain/Rank.java:
```diff
@@ -0,0 +1,34 @@
+                .filter(rank -> rank.matchCount == matchCount)
+                .findFirst()
+                .orElse(NONE);
+    }
```
지윤님이 생각하시는 of와 from을 사용하는 기준이 있을까요?

[정적 팩토리 메소드](https://tecoble.techcourse.co.kr/post/2020-05-26-static-factory-method/)에 대한 글을 하나 가져와봤습니다 :)

dd-jiyun replied:
지금까지는 `of`와 `from`을 비슷한 의미로 받아들여서, 딱히 구분하지 않고 사용해왔던 것 같아요.

그런데 공유해주신 글을 보고 나서, 정적 팩토리 메서드에도 네이밍 컨벤션이 있다는 걸 새롭게 알게 되었어요!

`of` → 여러 값을 조합해서 객체를 만들 때
`from` → 하나의 값에서 다른 객체로 변환할 때
`valueOf` → 문자열이나 숫자에서 Enum을 얻을 때

이 기준으로 보면, `Rank` 클래스에서 `matchCount` 값을 기반으로 Enum을 찾는 경우는 `valueOf()`가 가장 적절해 보이기도 합니다.

하지만, 이미 자바에서 제공하는 `Enum.valueOf(String name)` 메서드가 있어서, 매개변수 타입이 다르더라도 `동일한 메서드명을 사용하면 혼란이 생기지 않을까?` 하는 고민이 들었어요.
그래서 저는 이런 경우 `from()`을 사용하는 게 더 안전한 선택일 수 있다고 생각했습니다.

혹시 주노는 이런 상황에서 어떤 기준으로 메서드 이름을 선택하시나요?
또, 실무에서 이러한 네이밍이 컨벤션과 다를 경우 실제로 혼동을 준 경험이 있으셨는지도 궁금합니다!

Choi-JJunho replied:
말씀해주신 대로 valueOf의 경우 Enum에서 선언되어있는 메소드라서 저는 사용을 지양하는 편이에요!
개인적으로는 파라미터가 1개일때는 from, 2개 이상일때는 of 를 사용하는 정도로만 해당 컨벤션을 활용하고 있습니다 :)

Choi-JJunho commented on src/main/java/domain/WinningLotto.java:
```diff
@@ -0,0 +1,44 @@
+
+    private List<Integer> parseToInt(final String input) {
+        try {
+            return Arrays.stream(input.split(SPLIT_DELIMITER))
+                    .map(String::strip)
+                    .map(Integer::parseInt)
+                    .toList();
```
stream 내부에서 예외가 발생하면 stream 전체가 중단되는데요!
for문과 stream 각각의 내부에서 예외가 발생하는것에는 어떤 차이가 있는지 알아보면 좋을 것 같아요!

(현재 로직에 고쳐야하는부분이나 문제가 있다는것은 아닙니다!)

dd-jiyun replied:
둘의 차이를 찾아보니, for문은 외부 반복자, stream은 내부 반복자라는 개념으로 구분할 수 있었습니다!

외부 반복자인 `for`문은 반복 흐름에 대한 제어권이 개발자에게 있어서 continue, break, return 등을 자유롭게 사용할 수 있고, 예외가 발생해도 그 이후 로직을 세밀하게 제어하면서 계속 진행할 수 있는 장점이 있더라고요.

하지만, `stream`은 내부 반복자라서 반복의 흐름이 라이브러리에 위임되어 있어서 예외 발생 시 즉시 종료되어 버립니다. 
`try-catch`로 처리할 수는 있지만, 반복 구조 안에 예외 처리를 넣으면 오히려 가독성이 떨어지고 복잡해지는 문제도 생기더라고요.

그래서 상황에 따라 for문과 stream을 예외 처리와 반복 흐름 제어라는 관점에서 구분해서 선택해야겠다는 생각이 들었습니다 👍🏻 

[참고한 글 : stream vs. for-loop](https://www.baeldung.com/java-streams-vs-loops)

Choi-JJunho commented on src/main/java/domain/WinningStatistics.java:
```diff
@@ -0,0 +1,35 @@
+public class WinningStatistics {
+
+    private final WinningLotto winningLotto;
+    private final Map<Rank, Integer> rankCounts = new HashMap<>();
```
Key가 Enum인 Map이군요! EnumMap을 사용해보면 어떨까요?

dd-jiyun replied:
Enum으로 key할거면 확실히 EnumMap을 사용하는게 좋네요!
HashMap과 비교해보면서 EnumMap이 순서를 보장하고 성능도 더 좋다는 장점이 있다는 걸 알게되었어요.
지금처럼 enum 정의 순서를 변경할 필요가 없는 상황에서는 EnumMap을 사용하는 게 더 나은 선택인 것 같아요 👍🏻

Choi-JJunho commented on src/main/java/domain/WinningStatistics.java:
```diff
@@ -0,0 +1,35 @@
+    public long getTotalPrize() {
+        long total = 0;
+        for (Rank rank : Rank.values()) {
+            total += (long) rank.getPrize() * getCount(rank);
```
rank의 prize가 int인 이유가 있을까요? (22억이 넘는 금액이 Prize가 되면 어떻게 될까요?)

dd-jiyun replied:
헉 22억까지는 생각하지 않아서 int로 선언했어요,,
하지만, 금액과 관련된 값은 언제든 변경될 수 있다는 점을 생각해보면 long으로 받는 게 더 안정적일 것 같아요 !

Choi-JJunho replied:
한발 더 나아가서 금액은 무한히 늘어날 수 있고, 값이 틀어질때 매우 치명적인 문제를 야기할 수 있는 도메인인점을 감안하면 long의 최댓값도 결국 안전하지 못한 상황이 올텐데요!

Money라는 도메인을 안전하게 다루려면 어떤 방법을 사용해볼 수 있을까요?

> 키워드 : BigDecimal, 문자열

dd-jiyun replied:
`BigDecimal`이 `long`보다 금액 도메인을 다룰 때 훨씬 안정적이고 정확하게 처리할 수 있다는 점을 알게 되었어요. 
그래서 리뷰에 나온 키워드를 참고해 `Prize` 클래스를 만들어 당첨 금액 계산에 long 대신 BigDecimal을 사용하도록 개선했습니다!

추가로 저는 "문자열"이라는 키워드를 구매 금액을 사용자에게 문자열로 입력받는 부분이라고 이해했어요. 그래서 `PurchaseAmount`에서 문자열을 `BigDecimal`로 변환하는 방식으로 처리했는데 수정하면서 오류가 많이 발생했고, 또 예외처리를 Application.main에서 담당하게 되더라고요.

이 부분에서 아직 개념이 완전히 잡히지 않은 것 같아요 😂
입력을 문자열로 받고, 도메인 객체로 변환하는 흐름과 예외처리 책임을 어디까지 분리해야 하는지 고민 중이에요..

제가 제대로 이해하고 있는게 맞는지 조금 더 고민을 해봐야할 것 같습니다,,,,

Choi-JJunho replied:
저도 답변드리다가 찾게된 내용인데 BigDecimal 이외에도 BigInteger라는 친구도 있었네요!
Money VO로 감싸보라는 의도로 말씀드렸던 내용이였는데 꽤나 많은 길을 헤매게 만든것 같네요 😂

다음 단계에서 이어서고민해보시죠!

Choi-JJunho commented on src/test/java/domain/WinningLottoTest.java:
```diff
@@ -0,0 +1,88 @@
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+class WinningLottoTest {
```
꼼꼼한 테스트 좋네요! 👍

Choi-JJunho commented on src/test/java/domain/WinningStatisticsTest.java:
```diff
@@ -0,0 +1,54 @@
+        long totalPrize = statistics.getTotalPrize();
+
+        // then
+        assertThat(totalPrize).isEqualTo(2_000_000_000L + 1_500_000L);
```
상수로 테스트를 구성해보는건 어떨까요?

`Rank.FIRST.getPrize()` ...

dd-jiyun replied:
오! 당첨 금액이 바뀌면 이 상수도 바꿔야하는 불편함이 있을 것 같네요! 
좋은 조언 감사합니다 👍🏻

Choi-JJunho reviewed (commented):
고생하셨습니다~!
코멘트 몇개 남겼으니 확인 부탁드릴게요

izzy80 commented:
2단계 미션 수고 많으셨어요~ 지윤님 코드는 항상 깔끔해서 참고가 많이 됩니다!
로또 등수별 상금 정보를 어떻게 관리할지 고민했는데, enum과 hashmap 조합으로 풀어낸 방식이 깔끔하네요! 
더 객체지향적인 것 같고!!!
저는 처음에 객체로 풀어야 하나 고민했었는데, 많이 배워갑니다!

izzy80 commented on src/main/java/view/OutputView.java:
```diff
@@ -2,19 +2,22 @@
 
 public class OutputView {
 
     private OutputView() {
```
모든 메서드가 static으로 구성된 OutputView에서 기본 생성자를 private으로 막아두시는 군요!!
실수로 객체를 생성하지 못하게 방지하려는 의도가 느껴집니다 ^0^

izzy80 commented on src/test/java/domain/ProfitRateCalculatorTest.java:
```diff
@@ -0,0 +1,35 @@
+    @Test
+    @DisplayName("정상적으로 수익률을 계산한다.")
+    void shouldReturnCalculateProfitRate() {
+        // given
```
given / when / then 구조가 깔끔하게 잘 드러나 있어서 테스트 흐름이 한눈에 들어오네요! 🙌

dd-jiyun committed:
refactor: 정적 팩토리 메서드 네이밍 컨벤션에 맞게 수정.

- LottoNumber의 생성자를 제거하여 정적 팩토리 메서드로 객체를 생성할 수 있게 변경하였습니다.

dd-jiyun reviewed (commented):
세세한 리뷰 감사합니다 👍🏻

dd-jiyun committed:
refactor: HashMap에서 EnumMap으로 수정

dd-jiyun committed:
refactor: prize 값을 long으로 변환

Choi-JJunho commented on src/main/java/domain/WinningLotto.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class WinningLotto {
```
생각해볼만한 부분이 될 것 같아 남겨봅니다!
WinningLotto가 Lotto를 상속받는 구조에 대해 어떻게 생각하시나요?

상속은 언제 사용했을 때 효과가 있다고 생각하시나요?

dd-jiyun replied:
음, 주노의 말씀을 듣고 제가 어떤 의도로 설계했는지 다시 돌아보게 되었어요!

`WinningLotto`를 만들면서, 당첨 번호 입력 검증 로직이 `Lotto`와 거의 동일해서 검증 책임을 중복하지 않으려고 내부에서 `new Lotto()`를 통해 처리하도록 했습니다.

하지만 `WinningLotto`가 `Lotto`의 모든 행위와 속성을 그대로 가져야 한다기보다는, "당첨 번호"라는 특별한 역할이 있기 때문에 상속보다는 포함하는 편이 더 적절하다고 생각했어요.

상속은 '~은 ~이다' 관계가 명확할 때 효과적이라고 알고있는데요! 예를 들어 `강아지는 동물이다`처럼, 상위 개념의 속성과 기능을 그대로 물려받고 거기에 `강아지만의 특징을 추가하는 경우`에 잘 맞는다고 생각합니다.

`WinningLotto`와 `Lotto`는 `당첨 번호 로또`와 `일반 로또`라는 차이가 있어서, 둘이 완전히 같은 종류는 아니라고 생각해서 상속보다는 포함 관계가 더 적합하다고 판단했습니다.

그래서 상속은 `같은 종류면서 기능을 확장 또는 세분화할 때` 그리고 `코드 중복을 줄이기 위해 공통 기능을 공유할 때`가 가장 효과적이라고 생각합니다. 😆

dd-jiyun committed:
refactor: 당첨 금액을 감싸는 Prize 클래스 생성 및 수익률 계산을 통계 클래스로 이동

- 통계하는 클래스에서 수익률 계산까지 하는 것이 자연스러운 흐름이라고 판단했습니다.
- long보다 더 큰 당첨 금액의 경우의 문제를 방지하고자 BigDecimal로 변경하였습니다.

Choi-JJunho commented on src/main/java/domain/Prize.java:
```diff
@@ -0,0 +1,49 @@
+
+    private Prize(final BigDecimal amount) {
+        validateNegativeNumber(amount);
+        this.amount = amount.setScale(DECIMAL_PLACES, RoundingMode.HALF_UP);
```
setScale도 설정해주셨군요! 👍

Choi-JJunho commented on src/main/java/domain/PurchaseAmount.java:
```diff
@@ -1,18 +1,11 @@
 package domain;
 
-public class PurchaseAmount {
+public record PurchaseAmount(int amount) {
```
여기에서 다루는 amount는 int네요!
돈과 관련된 모든 값을 래핑해볼 수 있을까요?

(다음 미션에서 진행해보시죠 ㅎㅎ)

dd-jiyun replied:
네!! 조언 감사합니다 👍

jisoo78 commented:
지윤님 이번 미션 고생 많았습니다!
지윤님은 이번에 바로 MVC 패턴을 적용하고 진행하셨네요...! 
저도 보고 배워야겠습니다
처음 달아보는 리뷰라 많이 어색하지만 열심히 달아보겠습니다!

jisoo78 commented:
지윤님 이번 미션 고생 많았습니다!
이전 미션에 적용한 MVC 패턴을 바로 적용하셨네요...!
저도 보고 배워야겠네요🥑
처음 해보는 리뷰라 많이 어색하지만 열심히 달아보겠습니다!
